Support for pickling sentinel objects as singletons

HexDecimal opened the pull request:
My attempt at implementing [PEP 661](https://peps.python.org/pep-0661/) since I was unhappy with #594. I'm hoping that this is a PR of decent quality and not just me desperately wanting pickle support.

Changes made to Sentinel:

- ~~Added `module_name` parameter following PEP implementation and tweaking that to use the local `_caller` helper function. Breaking change due to `repr` becoming a keyword parameter.~~ `module_name` removed due to changes in PEP 661.
- `name` requires qualified names to support pickle singletons, this is tested and documented.
- Added `__reduce__` to track Sentinel by name and module_name.
- Added copy and pickle tests. There were several other use-cases for pickle from the PEP 661 discussion. The PEP requires that the sentinel identity is preserved during these operations.
- Updated documentation for Sentinel.

For a while I still supported the `repr` parameter and after following the [PEP 661 discussion](https://discuss.python.org/t/pep-661-sentinel-values/9126) I ended up implementing the truthiness tweaks mentioned there, but then I ended up scrapping all of that so that I could follow the PEP more closely, but they would be easy to reintroduce later if desired.

I'm unsure of how to mention version changes in the docs. Replacing `repr` with `module_name` is technically a breaking change.

Closes #720
Closes #742

Viicos commented:
> Replacing `repr` with `module_name` is technically a breaking change.

I think breaking changes are expected for draft PEPs, so imo no deprecation process should be used. Users should be aware that changes are expected. However, I think we can:
- Improve [the documentation about draft PEPs](https://github.com/python/typing_extensions/blob/59d2c20858ac527516ebad5a89c05af514dac94a/doc/index.rst?plain=1#L16-L21), stating that implementation can change.
- Explicitly state that Sentinels are from a draft PEP in the `Sentinel` class documentation, and isn't stable yet.

HexDecimal commented:
> I'm not willing to remove the `repr` parameter from `typing_extensions.Sentinel` without a deprecation. However, I'm OK to re-export the builtin Sentinel class in 3.15 (if it makes it in), even if it doesn't have this parameter.

Then I will change `repr` into a keyword parameter.

If compatibility is important enough then I can also have the code assume `module_name` is actually `repr` if it does not refer to a existing module. Would this be necessary?

Configuration like `repr` would be defined only with the initial Sentinel object. It wouldn't get stored in the reduce function. This allows changes in `repr` to be reflected in unpickled sentinels. This means that removing `repr` later will not break serialization.

HexDecimal commented:
I've replaced the `__reduce__` method with a version using pickle's singleton support. This is the most conservative and inoffensive option for a new reduce function since it doesn't add a custom unpickle function, is forward compatible with any future method of pickling, is the standard method of handling singletons via pickle, and is generally strict. This does not support pickling anonymous sentinels which will now raise `pickle.PicklingError` instead of `TypeError`. I can revert this if the previous behavior was more desired, but I personally only need to pickle sentinels which have a top-level definition, and the more I work with them the more that anonymous sentinels feel counter intuitive.

I've attempted to add to the discussion of PEP 661. Anonymous sentinels bring a lot of issues which need to be addressed.
https://discuss.python.org/t/pep-661-sentinel-values/9126/251

Before this is merged I'd like to add a `strict` keyword parameter to Sentinel which prevents sentinels from being accidentally named after existing top-level objects which are not Sentinel as well as ensuring that `repr` is never given conflicting values. I'll need feedback on if this is appropriate.

Viicos commented:
Thanks @HexDecimal. I just tested your implementation in Pydantic to implement an `UNSET` sentinel, and [pickling seems to work as expected](https://github.com/pydantic/pydantic/pull/11883/files#diff-b446ead7cd1bc63e9765e6faf001b14ea47dc50c2c33ea4fa7965fe548b24d2fR50-R58).

HexDecimal commented on src/typing_extensions.py:
```diff
@@ -4155,24 +4156,87 @@ def evaluate_forward_ref(
+            return None
+        try:
+            module = importlib.import_module(module_name)
+            return operator.attrgetter(name)(module)
```
This suggestion breaks qualified names and current tests. `operator.attrgetter` handles dots while `getattr` does not.

HexDecimal replied:
Relevant code removed during refactor.

HexDecimal commented:
Note on behavior regarding the sentinel registry and imported sentinels:
```py
class MISSING:
    pass

assert Sentinel("MISSING") is MISSING  # Sentinel named after an existing non-Sentinel object
```
Which of these makes the most sense for the above code?

1. Register and return the existing class as the sentinel identity, the assert will pass (current PR implementation)
2. Return a new Sentinel identity, the assert will fail (old implementation)
3. Raise a warning, then do 1
4. Raise `TypeError` because `MISSING` is not a `Sentinel` (what I recommend, the most strict option)
5. Add a `strict` parameter to `Sentinel` to decide between 1 and 4

---

2nd behavior involving registered sentinel parameters:

```py
MISSING = Sentinel("MISSING", repr="foo")

assert MISSING is Sentinel("MISSING")  # Implicit 'repr' conflicts with previous 'repr'
assert MISSING is Sentinel("MISSING", repr="bar")  # Explicit 'repr' conflicts with previous 'repr'
```
Which of these makes the most sense for the above code?

1. `repr="foo"` is locked in for all subsequent sentinels, subsequent `repr` is always ignored, asserts pass (current implementation)
2. Both asserts raise `ValueError` as long as `repr="foo"` is not given again (my recommendation, the most strict option)
3. `repr` only checked if explicitly given, 1st assert passes, 2nd assert raises `ValueError`
4. Check `repr` like 2 but raise warnings instead of errors
5. Add `strict` parameter to decide between 1 and 2

---

The truthiness of sentinels is still being discussed. One thing which is unclear is if conversions to bool are common enough to be an issue in the first place (because sentinels are normally checked using identity comparison and little else). It'd be interesting to experiment with the most strict option and disable conversions to bool entirely:

```py
def __bool__(self) -> Never:
    raise TypeError(...)
```

HexDecimal commented:
> Constructing a class shouldn't look around in the globals for other stuff that might be using the same name.

That was necessary to handle unpickling until I finally implemented the correct reduce function, but at this point I can remove that code now and revert to the old behavior. I'd accept that the other options are unnecessarily handholdy.

HexDecimal committed:
Support for pickling sentinel objects as singletons

The `repr` parameters position was replaced by `module_name` to conform to PEP 661.

Added copy and pickle tests.

Updated documentation for Sentinel.

`_marker` was defined before `caller` which causes minor issues,
resolved by setting its module name manually.

HexDecimal commented:
I've narrowed the scope of this PR down to only what's needed to support pickling sentinels. Unfortunately this still means that `repr` needs to be touched to allow for the `module_name` parameter. Thankfully a lot of other changes were no longer needed after this latest rebase.

Has anyone mentioned that `_marker` is a terribly obfuscated name for a sentinel?

HexDecimal committed:
Change module_name into a keyword-only parameter

HexDecimal commented on src/typing_extensions.py:
```diff
@@ -164,18 +164,26 @@ class Sentinel:
     def __init__(
         self,
         name: str,
+        module_name: typing.Optional[str] = None,
```
`module_name` is now keyword-only.

JelleZijlstra commented:
This is mostly good but we shouldn't add a module_name argument since PEP-661 doesn't have that; see #742.

@HexDecimal thanks for the persistent work here! Are you interested in the remaining work of making Sentinel forward-compatible with the accepted version of PEP 661, as outlined in #742?

HexDecimal commented:
I will work on the PR right now.

HexDecimal committed:
Move _caller above sentinel

Sentinels _marker requires _caller in order to remove its module_name

HexDecimal committed:
Update sentinel for PEP 661 changes

Rename Sentinel to sentinel, deprecated old name

Remove module_name parameter

Deprecate subclassing sentinel

Enforce correct sentinel parameters using deprecated overloads

Rename _name attribute to __name__

Rename sentinel in tests, tests passed before making this change

Also add tests for sentinel deprecations

HexDecimal committed:
Deprecate sentinel repr parameter

`repr` is not in PEP 661

HexDecimal commented:
I've had to move `overload` and `deprecated` in order to perform high-quality deprecations of older sentinel features. I've also had to move `_caller` in order to remove `module_name` from sentinels.

HexDecimal commented:
I'm having trouble debugging this error: https://github.com/python/typing_extensions/actions/runs/24916592995/job/72969774720?pr=617
Any advice for this? Only affects Python 3.11.15. Deleting doctests affected this this PR does not resolve it.

HexDecimal committed:
Skip deprecation of Sentinel class

HexDecimal commented:
I attempted to deprecate the `Sentinel` class using a `@deprecated` decorator but this broke Python 3.11.15 specifically, not 3.11.0 or 3.12.x or any other version.

HexDecimal commented:
* Renamed `Sentinel` to `sentinel`, Old name kept, I wanted to deprecate the old name [but was prevented due to this error in Python 3.11.15](https://github.com/python/typing_extensions/actions/runs/24916592995/job/72969774720?pr=617), this specific version does not handle PEP 702 deprecation of classes.
* Deprecated the `repr` parameter and removed it from the documentation. Parameter deprecations are enforced via PEP 702 style deprecations, including enforcing `name` as a positional-only parameter.
* Deprecated subclassing of `sentinel`
* Renamed `_name` attribute to `__name__`
* Strongly discouraged assigning attributes or using `__weakref__` via the documentation. I'm assuming this will switch to `__slots__` later.

Hopefully the code quality here is acceptable. Skipping deprecations would've cleaned this up a lot. Accounting for PEP 702 ahead of time would've prevented code from being shuffled around.

JelleZijlstra commented on doc/index.rst:
```diff
@@ -1071,18 +1071,20 @@ Capsule objects
 ~~~~~~~~~~~~~~~~
 
-.. class:: Sentinel(name, repr=None)
+.. class:: sentinel(name, /)
```
We should keep it named as Sentinel, I don't think it's worth trying to rename it. We'll just have the slightly odd case that `typing_extensions.Sentinel == builtins.sentinel`.

HexDecimal replied:
Using the correct name is important. Switching from `typing_extensions` to native Python and backporting to older versions should be a simple addition or removal of `from typing_extensions import sentinel`. Using Sentinel with the wrong case needs to be actively discouraged or else things will get worse for code migration.

JelleZijlstra replied:
Fair, I'd like to hear more opinions on this though. I'll post in the issue.

JelleZijlstra commented on doc/index.rst:
Add a versionchanged 4.16.0 with a summary of the changes in this version.

JelleZijlstra commented on doc/index.rst:
```diff
@@ -1091,6 +1093,13 @@ Sentinel objects
       ...
       >>> func(MISSING)
 
+   Sentinels defined inside a class scope should use a :term:`qualified name`.
```
I'm not sure we need to mention this, it's a fairly unusual use case and typing-extensions docs don't need to be very precise; we can refer to the CPython docs for more.

HexDecimal replied:
That makes sense. The reference to PEP 661 should be enough. Anybody else will learn about the qualified name requirement when they try to use pickle.

JelleZijlstra commented on src/typing_extensions.py:
```diff
@@ -159,22 +160,270 @@
-    If not provided, "<name>" will be used.
     """
 
+    @overload
```
I'm not sure the overloads are worth it, since type checkers won't look at the source for this package. It's enough to raise a runtime DeprecationWarning, and we can attempt type checker deprecations in typeshed.

I think we can write the signature something like this:

```python
def __init__(self, _name: str = _marker, /, repr: Optional[str] = None, *, name: str = _marker) -> None:
    if name is not _marker:
       warnings.warn(DeprecationWarning, "Passing `name` as a keyword argument is deprecated")
       _name = name
     if _name is _marker:
         raise TypeError("name is required")
     if repr is not None:
         warnings.warn(DeprecationWarning, ...)
```

And in typeshed we can use overload+deprecated as you suggest.

HexDecimal replied:
I think you're right. I thought PEP 702 would look more clean, and it does, but this is useless outside of typeshed.

I'll revert my changes and switch to this.

HexDecimal replied:
Note that sentinels can not be used as defaults for the sentinels own methods. This all happens at import-time.

JelleZijlstra replied:
True, good point. We'll have to use `object()`.

HexDecimal committed:
Switch to manual warnings for outdated sentinel parameters

HexDecimal committed:
Switch to manual warning of sentinel subclassing

HexDecimal committed:
Note user-facing changes to sentinel

Remove redundant info about sentinels, specifics from PEP 661 do not need to be repeated.

JelleZijlstra commented on doc/index.rst:
```diff
@@ -1095,6 +1094,15 @@ Sentinel objects
 
+   .. versionchanged:: 4.16.0
+
+      Now supports pickle and will be reduced as a singleton.
```
Add a line: "The implementation of this class was updated to conform to the accepted version of :pep:`661`."

HexDecimal replied:
Added along with your other suggestions.

JelleZijlstra commented on src/typing_extensions.py:
```diff
@@ -159,22 +160,75 @@
+_sentinel_placeholder = object()
+
+
+class sentinel:
```
Can you make this re-export builtins.sentinel on 3.15, so something like `if hasattr(builtins, "sentinel"): sentinel = builtins.sentinel`

HexDecimal replied:
Thanks. I was wondering when I was supposed to add this.

`hasattr(builtins, "sentinel")` is correct. I can't imagine any other method being reliable.

JelleZijlstra commented on src/typing_extensions.py:
```diff
@@ -159,22 +160,75 @@
+    ) -> None:
+        if name is not _sentinel_placeholder:
+            warnings.warn(
+                "'name' is positional-only and must not be a keyword parameter",
```
```suggestion
                "Passing 'name' as a keyword argument is deprecated; pass it positionally instead.",
```

JelleZijlstra commented on src/typing_extensions.py:
```diff
@@ -159,22 +160,75 @@
+            raise TypeError("First parameter 'name' is required")
+        if repr is not None:
+            warnings.warn(
+                "'repr' is deprecated and must be removed",
```
```suggestion
                "The 'repr' parameter is deprecated and will be removed in Python 3.15.",
```

JelleZijlstra commented on src/typing_extensions.py:
```diff
@@ -159,22 +160,75 @@
+
+    def __init_subclass__(cls):
+        warnings.warn(
+            "Subclassing sentinel is forbidden by PEP 661",
```
```suggestion
            "Subclassing sentinel is deprecated and will be disallowed in Python 3.15",
```

JelleZijlstra commented on src/typing_extensions.py:
```diff
@@ -159,22 +160,75 @@
+    def __setattr__(self, attr: str, value: object) -> None:
+        if attr not in {"__name__", "_repr", "__module__"}:
+            warnings.warn(
+                "Setting attributes on sentinel is deprecated",
```
```suggestion
                f"Setting attribute {attr!r} on sentinel objects is deprecated and will be disallowed in Python 3.15.",
```

JelleZijlstra commented on doc/index.rst:
```diff
@@ -1071,18 +1071,17 @@ Capsule objects
 
-   If *repr* is provided, it will be used for the :meth:`~object.__repr__`
-   of the sentinel object. If not provided, ``"<name>"`` will be used.
+   Assigning attributes to a sentinel including `__weakref__` is forbidden.
```
```suggestion
   Assigning attributes to a sentinel is deprecated.
```

JelleZijlstra commented on doc/index.rst:
```diff
@@ -1095,6 +1094,15 @@ Sentinel objects
+      Automatic `repr` string no longer has angle brackets.
+      `repr` parameter was deprecated.
+      `name` as a keyword is deprecated.
+      Subclasssing and attribute assignment are deprecated.
```
```suggestion
      Subclassing and attribute assignment are deprecated.
```

HexDecimal committed:
Apply suggestions from code review

Update related tests

Co-authored-by: Jelle Zijlstra <jelle.zijlstra@gmail.com>

HexDecimal committed:
Port sentinel from Python when available
